Infrastructure/ClassroomRepository.cs

```diff
@@ -273,9 +268,34 @@ UPDATE session_in_classroom
         return Result.Ok();
     }
 
-    public async Task<Result> JoinClassroomAsync(int studentId, int classroomId)
+    public async Task<Result> JoinClassroomAsync(int studentId, string roomCode)
     {
         using var con = await _connection.CreateConnectionAsync();
+
+        var getClassroomIdQuery = "SELECT classroom_id FROM classroom WHERE roomcode = @RoomCode";
+
+        var classroomId = await con.QueryFirstOrDefaultAsync<int>(getClassroomIdQuery, new { RoomCode = roomCode });
+        if (classroomId == 0)
+        {
+            _logger.LogInformation("User {userID} tried to join classroom with invalid code {code}", studentId, roomCode);
+            return Result.Fail("Invalid roomcode");
+        }
+
+        var registrationOpen = await VerifyRegistrationIsOpen(classroomId);
+        if (!registrationOpen)
+        {
+            _logger.LogInformation("User {userID} failed to join classroom {classroomID} - registration not open", studentId, classroomId);
+            return Result.Fail("Classroom cannot not open to join");
+        }
+
+        var checkJoinedQuery = "SELECT 1 FROM student_in_classroom WHERE student_id = @StudentId AND classroom_id = @ClassroomId";
+        var joinedAlready = await con.QuerySingleOrDefaultAsync<int>(checkJoinedQuery, new { StudentId = studentId, ClassroomId = classroomId });
+        if (joinedAlready != 0)
+        {
+            _logger.LogInformation("User {userID} tried to join classroom {classroomId}, but was already joined", studentId, classroomId);
+            return Result.Fail("Student already joined classroom");
+        }
+
         var transaction = con.BeginTransaction();
```

Comment:
When we create the checks above shouldnt we ensure that they also run in the transaction such that we know they have correct data available.